### PR TITLE
F53 functionality for duro aaeon

### PR DESCRIPTION
**Pull request by kokipsiho**

### CODE REVIEW

**Comment by joshmh** [comment]
[F53 Functionality for Douro/Aaeon.](https://trello.com/c/D16ehMh8/585-f53-functionality-for-douro-aaeon)

File: lib/brain.js [hunk 1]
```diff
@@ -2601,84 +2601,160 @@ Brain.prototype._dispense = function _dispense () {
...
+  
+  function* dispenseGenerator() {
+    let result = []
+    for (let index = 0; index < notesToDispense.length; index++) { // cannot yield inside forEach
+      let yielded = yield this.billDispenser.dispense(notesToDispense[index])
+      if (yielded !== undefined) result.push(yielded)
+      batchNum = index + 1
+      self.browser().send({dispenseBatch: {current: batchNum, of: notesToDispense.length} })
+    }
+    yield Promise.all(result.map(i => i.value))
```
Comment (joshmh) [inline lib/brain.js]: Does this have to be a promise? Isn't this just an array of numbers?

File: lib/brain.js [hunk 2]
```diff
@@ -2601,84 +2601,160 @@ Brain.prototype._dispense = function _dispense () {
...
+      .then(result => {
+        const bills = result.map(d => d.bills).reduce((p, c) => {
+          return [
```
Comment (joshmh) [inline lib/brain.js]: Need a unit test for this and for the ``optimize`` function.

File: lib/brain.js [hunk 3]
```diff
@@ -2601,84 +2601,160 @@ Brain.prototype._dispense = function _dispense () {
...
+          optimizedArray.pop()
+          return
+        }
+      findLimitSum(a - head, b - tail, limit)
+    }
+  
+    findLimitSum(inputHead, inputTail)
+
+    return optimizedArray
+  }
```
[the remark this answers is not in the document]
Comment (kokipsiho) [reply lib/brain.js]: This code returns only one element (array) in array while my implementation returns array of arrays. Do you suggest this to be used in some loop or other part of the code?

I've tested this with a lot of cases but you are right, I missed one and first dispense can be less than limit.

File: lib/brain.js [hunk 4]
```diff
@@ -2601,84 +2601,160 @@ Brain.prototype._dispense = function _dispense () {
...
+Brain.prototype.waitForCollection = function waitForCollection (dispenseGenerator) {
+  const txId = this.tx.id
+  let g = dispenseGenerator.call(this) // change execution context with .call
+  let start = true
+  let yieldedVal
+
+  const checkCollection = () => {
+    const isCollecting = this.tx.id === txId && this.state === 'dispensing'
+
+    if (start) yieldedVal = g.next()
```
[the remark this answers is not in the document]
Comment (kokipsiho) [reply lib/brain.js]: That's correct. Tested and working.

File: lib/brain.js [hunk 5]
```diff
@@ -2601,84 +2601,160 @@ Brain.prototype._dispense = function _dispense () {
...
+        }
+      findLimitSum(a - head, b - tail, limit)
+    }
+  
+    findLimitSum(inputHead, inputTail)
+
+    return optimizedArray
+  }
+
+  notesToDispense = optimize(notes, this.billDispenser.dispenseLimit)
```
[the remark this answers is not in the document]
Comment (kokipsiho) [reply lib/brain.js]: Agreed

File: lib/brain.js [hunk 6]
```diff
@@ -2601,84 +2601,160 @@ Brain.prototype._dispense = function _dispense () {
     })
 }
 
+Brain.prototype.waitForCollection = function waitForCollection (dispenseGenerator) {
+  const txId = this.tx.id
+  let g = dispenseGenerator.call(this) // change execution context with .call
+  let start = true
+  let yieldedVal
```
[the remark this answers is not in the document]
Comment (kokipsiho) [reply lib/brain.js]: All vars are used in function they are declared, `start` will be removed in next commit.

File: lib/brain.js [hunk 7]
```diff
@@ -2601,84 +2601,160 @@ Brain.prototype._dispense = function _dispense () {
...
   this._transitionState('dispensing')
   emit('billDispenserDispensing')
+  
+  function* dispenseGenerator() {
+    let result = []
+    for (let index = 0; index < notesToDispense.length; index++) { // cannot yield inside forEach
+      let yielded = yield this.billDispenser.dispense(notesToDispense[index])
+      if (yielded !== undefined) result.push(yielded)
+      batchNum = index + 1
+      self.browser().send({dispenseBatch: {current: batchNum, of: notesToDispense.length} })
```
[the remark this answers is not in the document]
Comment (kokipsiho) [reply lib/brain.js]: Code on line 2679 is setting UI before transitioning to dispense screen while code in loop is incrementing current batch

File: lib/brain.js [hunk 8]
```diff
@@ -2601,84 +2601,160 @@ Brain.prototype._dispense = function _dispense () {
...
     console.log('Wrong dispenser currency; dispenser: %s, tx: %s',
       this.billDispenser.fiatCode, fiatCode)
     return this._timedState('wrongDispenserCurrency')
   }
 
+  self.browser().send({dispenseBatch: {current: batchNum, of: notesToDispense.length} })
   this._transitionState('dispensing')
   emit('billDispenserDispensing')
+  
+  function* dispenseGenerator() {
```
[the remark this answers is not in the document]
Comment (kokipsiho) [reply lib/brain.js]: Ok

File: lib/brain.js (hunk 1 above)
Comment (kokipsiho) [reply lib/brain.js]: It's an array of arrays, I used Promises to resolve those arrays into promise chain of previous implementation. Idea was to minimize change to current logic.

File: lib/brain.js (hunk 2 above)
Comment (kokipsiho) [reply lib/brain.js]: Got it

File: lib/brain.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshmh) [reply lib/brain.js]: It should return an array of arrays, try ``console.log(optimize([100, 55], 20))``. There's a bug though when processing an null limit, it just returns a flat array.

File: lib/brain.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (joshmh) [reply lib/brain.js]: Ah, ok.

File: lib/brain.js [hunk 9]
```diff
@@ -22,6 +22,7 @@ const usbreset = require('./usbreset')
 const version = require('../package.json').version
 const db = require('./db')
 const actionEmitter = require('./action-emitter')
+const optimizeDispense = require('./dispense-optimizer')
```
[the remark this answers is not in the document]
Comment (joshmh) [reply lib/brain.js]: 👍

File: lib/brain.js [hunk 10]
```diff
@@ -2630,131 +2627,102 @@ Brain.prototype.waitForCollection = function waitForCollection (dispenseGenerato
...
+Brain.prototype._dispenseGenerator = function* dispenseGenerator(notes, txId) { // TODO move to separate function
+  let batchNum = 1
+  let result = []
+  for (let index = 0; index < notes.length; index++) { // cannot yield inside forEach
+    let yielded = yield this.billDispenser.dispense(notes[index])
+    if (yielded !== undefined) result.push(yielded)
+    batchNum = index + 1
+    this.browser().send({dispenseBatch: {current: batchNum, of: notes.length} })
+  }
+  yield Promise.all(result.map(i => i.value))
```
[the remark this answers is not in the document]
Comment (joshmh) [reply lib/brain.js]: Right, I wouldn't initialize the whole brain object. It's probably easier to break this out into a separate file and make sure it doesn't use any brain stuff like ``this.billDispenser``. For instance, you could use the eventEmitter to send a message to brain to dispense. Then you could just look for the correct eventEmitter messages in the test.

[the remark this answers is not in the document]
Comment (kokipsiho) [reply lib/brain.js]: Got it, will do that tomorrow.

File: lib/dispense-generator.js [hunk 11]
```diff
@@ -92,4 +72,17 @@ function fullDispense (tx) {
...
 }
 
-module.exports = dispenseGenerator
\ No newline at end of file
+module.exports = function(notes, tx, txId) {
+  let g = dispenseGenerator(notes, tx, txId)
+  let yieldedVal = g.next()
+  
+  actionEmitter.on('billDispenser', event => g.next(event.value) )
+  actionEmitter.on('brain', event => {
```
Comment (kokipsiho) [inline lib/dispense-generator.js]: This will be removed, not used at all.

File: lib/dispense-generator.js [hunk 12]
```diff
@@ -1,80 +1,60 @@
...
+      !dispenseConfirmed
+        ? {
+          error: _.join(' ', _.reject(_.isEmpty, [result.name, result.message, result.err, result.error])),
+          errorCode: result.err,
+        }
+        : {})
+  })
+    
+  if (!dispenseConfirmed) {
+    console.log('dispense error5', result)
```
Comment (kokipsiho) [inline lib/dispense-generator.js]: I would throw actual error here

File: lib/brain.js [hunk 13]
```diff
@@ -466,6 +469,27 @@ Brain.prototype.processAction = function processAction (action) {
   }
 }
 
+Brain.prototype.processDispenseGeneratorAction = function processDispenseGeneratorAction (event) {
+  switch (event.action) {
+    case 'updateUI':
+      this.browser().send({dispenseBatch: {current: event.current, of: event.of} })
+      break;
```
Comment (joshmh) [inline lib/brain.js]: Install editor tool that enforces Standard format (no semicolons).

File: lib/brain.js [hunk 14]
```diff
@@ -466,6 +469,27 @@ Brain.prototype.processAction = function processAction (action) {
...
+    case 'dispenseBatch':
+      this.billDispenser.dispense(event.notes)
+      break;
+    case 'billDispenserDispensed':
+      break;
+    case 'billDispenserCollected':
+      break;
+    case 'fastUpdateTx':
+      return this.fastUpdateTx(event.value)
```
Comment (joshmh) [inline lib/brain.js]: Use either ``break`` or ``return`` throughout switch statement.

File: lib/dispense-generator.js [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  ]})
+
+  fillInBills(tx, bills)
+  const dispenseConfirmed = fullDispense(tx)
+
+  if (dispenseConfirmed) emit({action: 'billDispenserDispensed'})
+    // update tx and keep track of the
+    // dispensed bills
+    // and the error code (if any)
+  emit({
```
Comment (joshmh) [inline lib/dispense-generator.js]: I'd prefer if the emitted object was defined separately and broken down into smaller parts to make it easier to read.

File: lib/dispense-generator.js [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+  const total = _.sumBy(bill => bill.denomination * bill.dispensed, tx.bills)
+  return tx.fiat.eq(total)
+}
+
+module.exports = function(notes, tx, txId) {
+  let g = dispenseGenerator(notes, tx, txId)
+  let yieldedVal = g.next()
+  
+  actionEmitter.on('billDispenser', event => g.next(event.value) )
+}
```
Comment (joshmh) [inline lib/dispense-generator.js]: Install editor tool that enforces newline at end of file.

File: lib/process-door.js [hunk 17]
```diff
@@ -0,0 +1,19 @@
+const actionEmitter = require('./action-emitter')
```
Comment (joshmh) [inline lib/process-door.js]: This file (process-door.js) should be under ``lib/ssuboard``.

File: lib/ssuboard/led-manager.js [hunk 18]
```diff
@@ -81,11 +81,21 @@ function processBrain (event) {
...
+  }
+}
+
 function run () {
   console.log('DEBUG100')
   actionEmitter.on('fob', processFob)
   actionEmitter.on('door', processDoor)
   actionEmitter.on('doorManager', processDoorManager)
   actionEmitter.on('brain', processBrain)
+  actionEmitter.on('dispenseGenerator', processDispenseGenerator)
```
Comment (joshmh) [inline lib/ssuboard/led-manager.js]: Nice :)

**Comment by joshmh** [comment]
@kokipsiho I read over everything quickly and left some comments. Overall, I really like how things are broken out of brain.js and how that enables the unit tests. I'll try to find some time to look it over more carefully.

File: lib/mocks/billdispenser.js [hunk 19]
```diff
@@ -87,3 +94,9 @@ BillDispenser.prototype.dispense = function dispense (notes) {
       throw err
     })
 }
+
+BillDispenser.prototype.billsPresent = function billsPresent () {
+  return pDelay(2000).then(() => {
+    return false
+  })
```
[the remark this answers is not in the document]
Comment (kokipsiho) [reply lib/mocks/billdispenser.js]: stubFalse is nice trick, thx

**Pull request merged**